adsmp/app.py

```diff
@@ -294,110 +235,118 @@ def get_msg_status(self, msg):
-                            self.logger.error('Failed posting bibcode %s to Solr even without fulltext\nurls: %s, offending payload %s, error is  %s', failed_bibcode, solr_urls, doc, e)
+                            self.logger.exception('Failed posting bibcode %s to Solr even without fulltext (urls: %s)', failed_bibcode, solr_urls)
                             failed_bibcodes.append(failed_bibcode)
                     else:
                         # here if body not in error message do not retry, just note as a fail
                         self.logger.error('Failed posting individual bibcode %s to Solr\nurls: %s, offending payload %s, error is %s', failed_bibcode, solr_urls, doc, e)
                         failed_bibcodes.append(failed_bibcode)
-        return failed_bibcodes
+            # finally update postgres record
+            if failed_bibcodes and update_processed:
+                self.mark_processed(failed_bibcodes, checksums=None, type='solr', status='solr-failed')
 
-    
-    def mark_processed(self, bibcodes, type=None, status = None):
-        """Updates the timesstamp for all documents that match the bibcodes.
-        Optionally also sets the status (which says what actually happened 
-        with the document).
-        
+    def mark_processed(self, bibcodes, checksums=None, type=None, status=None):
+        """
+        Updates the timesstamp for all documents that match the bibcodes.
+        Optionally also sets the status (which says what actually happened
+        with the document) and the checksum.
+        Parameters bibcodes and checksums are expected to be lists of equal
+        size with correspondence one to one, except if checksum is None (in
+        this case, checksums are not updated).
         """
 
-        if not self.update_timestamps:
-            return
-        
         # avoid updating whole database (when the set is empty)
         if len(bibcodes) < 1:
             return
-        
-        if type == 'solr':
-            column = 'solr_processed'
-        elif type == 'metrics':
-            column = 'metrics_processed'
-        elif type == 'links':
-            column = 'datalinks_processed'
-        else:
-            column = 'processed'
-        
+
         now = adsputils.get_date()
-        updt = {column:now}
+        updt = {'processed': now}
         if status:
             updt['status'] = status
         self.logger.debug('Marking docs as processed: now=%s, num bibcodes=%s', now, len(bibcodes))
-        with self.session_scope() as session:
-            session.query(Records).filter(Records.bibcode.in_(bibcodes)).update(updt, synchronize_session=False)
+        if type is None:
+            timestamp_column = 'processed'
+            checksum_column = None
+            with self.session_scope() as session:
+                session.query(Records).filter(Records.bibcode.in_(bibcodes)).update(updt, synchronize_session=False)
+            session.commit()
```

Comment:
I'm afraid this is not what I meant, although it does clarify the code. 

Everywhere in the code we are now calling `mark_processed` with a specified `type`, I suggested we make `type` a mandatory argument:

```python
def mark_processed(self, bibcodes, type, checksums=None, status=None):
```

With only 3 accepted values ('solr', 'metrics' or 'links'), so we raise an Exception for any other value. And we **always** update the `processed` field independently of the type.

The `processed` field has an index in the DB and it is used by `run.py` to select/discard what records need to be indexed.